crates/dbsp/src/operator/dynamic/balance/accumulate_trace_balanced.rs

```diff
@@ -0,0 +1,1306 @@
+                        accumulator_snapshot_stream_val.clone(),
+                    );
+
+                    acc_feedback.connect(&accumulator_feedback_stream);
+
+                    // Exchange sender.
+                    let exchange_sender = RebalancingExchangeSender::<B, C>::new(
+                        batch_factories,
+                        worker_index,
+                        Some(location),
+                        exchange,
+                        accumulator_inner,
+                        balancer.clone(),
+                        circuit.metadata_exchange().clone(),
+                        self.local_node_id(),
+                    );
+
+                    let sender_node_id = circuit.add_ternary_sink(
+                        StreamingTernarySinkWrapper::new(exchange_sender),
+                        self,
+                        &delayed_acc,
+                        &delayed_trace,
+                    );
+
+                    // Exchange sender needs a persistent id.
+                    circuit.set_persistent_node_id(
+                        &sender_node_id,
+                        persistent_id
+                            .map(|persistent_id| {
+                                format!("{persistent_id}.balanced_exchange_sender")
+                            })
+                            .as_deref(),
+                    );
+
+                    // Add ExchangeSender -> ExchangeReceiver dependency.
+                    circuit.add_dependency(
+                        sender_node_id.local_node_id().unwrap(),
+                        sharded_stream.local_node_id(),
+                    );
+
+                    // Register integral with the balancer.
+                    balancer.register_integral(
+                        self.local_node_id(),
+                        sender_node_id.local_node_id().unwrap(),
+                    );
+
+                    (accumulator_stream, trace)
+                })
+            })
+            .clone()
+    }
+}
+
+/// The old contents of the integral and accumulator that needs to be rebalanced
+/// according to the new policy.
+struct RebalanceState<B: Batch<Time = ()>> {
+    /// Accumulator state doesn't need to be retracted, because it will be cleared when the rebalancing starts.
+    /// It only needs to be re-distributed using the new policy.
+    accumulator: SpineSnapshot<B>,
+
+    /// The policy that was used to accumulate the integral before the rebalancing.
+    /// This is the policy used during the previous transaction.
+    trace_policy: PartitioningPolicy,
+}
+
+/// Information about the key distribution of the input stream.
+// TODO: this can be refined in the future, e.g., it may be a useful to track heavy keys.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(transparent)]
+pub struct KeyDistribution {
+    /// Total weight of the data in each shard if we were using key-based sharding.
```

Comment:
hopefully there won't be records with very high weight
I wonder whether something heuristic could be done to reduce their contribution in case they do show up.

Reply:
right, I couldn't think of anything.